cog_layers/reader/io.py

```diff
@@ -0,0 +1,42 @@
+import os
+from dataclasses import dataclass
+import typing
+import functools
+
+import aiohttp
+import obstore as obs
+
+
+RangeRequestFuncType = typing.Callable[
+    [str, str, int, int, typing.Any], typing.Awaitable[bytes]
+]
```

Comment:
fwiw I'd highly recommend using class-based typing syntax here so we know what the parameters mean. You can define a protocol with a `__call__` method: https://github.com/developmentseed/obstore/blob/aaa8d6a0de646b532dc74a81f4adbd8ecf9de026/obstore/python/obstore/store/_aws.pyi#L644-L646

Reply:
Can you try this with your setup?
```py
from __future__ import annotations

from typing import Protocol


class Callback(Protocol):
    @staticmethod
    def __call__(a: int, b: str) -> bytes: ...


def requires_callback(callback: Callback) -> None:
    pass


def valid_callback(a: int, b: str) -> bytes:
    return b""


def invalid_callback1(a: int) -> bytes:
    return b""


def invalid_callback2(a: int, b: str) -> str:
    return ""


requires_callback(valid_callback)
requires_callback(invalid_callback1)
requires_callback(invalid_callback2)
```

<img width="781" alt="Screenshot 2025-02-28 at 3 43 55 PM" src="https://github.com/user-attachments/assets/825c01d4-cc74-472b-b677-4438489c699a" />

The last two lines fail the pylance type checker for me